src/main/java/sk/freemap/gpxAnimator/TileCache.java

```diff
@@ -0,0 +1,135 @@
+
+        //System.out.println("Caching enabled " + url);
+        return cachedGetTile( url, cachePath, cacheTimeLimit );
+    }
+
+    private static BufferedImage unCachedGetTile( String url ) throws UserException {
+        BufferedImage mapTile;
+        
+        System.setProperty("http.agent", "GPX Animator " + Constants.VERSION);
+        try {
+            mapTile = ImageIO.read(new URL(url));
+        } catch (final IOException e) {
+            throw new UserException("error reading tile " + url, e);
+        }
+        if (mapTile == null) {
+            throw new UserException("could not read tile " + url);
+        }
+        
+        return mapTile;
+    }
+    
+    private static BufferedImage cachedGetTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+        BufferedImage mapTile;
+        String fname = hashName(url);
+        String path = cachePath + File.separator + fname;
+        File f = new File(path);
+        
+        // Age out old file if needed
+        if (f.isFile()){
+            Date fileDate = new Date(f.lastModified());
+            long msBetweenDates = new Date().getTime() - fileDate.getTime();
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url); 
+                if (!f.delete()) {
+                    throw new UserException("failed to delete cache entry for " + url);
+                } 
+            }
+        }
+        
+        // If map tile is not in cache, then download it.
+        if (!f.isFile()) {
+            System.out.println("Cache miss for:" + path);
+            try {
+                FileUtils.copyURLToFile(new URL(url), f, 5000, 10000);
+            } catch (final IOException e) {
+                throw new UserException("error downloading tile " + url, e);
+            }
+        }
+        
+        // At this point a current map tile should be in our cache, so
+        // load it for the 
+        try {
+            mapTile = ImageIO.read(f);
+        } catch (final IOException e) {
+            throw new UserException("error reading tile " + url, e);
+        }
+        if (mapTile == null) {
+            throw new UserException("could not read tile " + url);
+        }
+        
+        return mapTile;
+    }
+    
+    private static String hashName(String url) throws UserException {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            return bytesToHex(digest.digest(url.getBytes(StandardCharsets.UTF_8)));
+        } catch (final NoSuchAlgorithmException e) {
+            throw new UserException("error creating hash name " + url, e);
+        }
+    }
+    
+    private static String bytesToHex(byte[] hash) {
+        StringBuffer hexString = new StringBuffer();
+        for (int i = 0; i < hash.length; i++) {
+        String hex = Integer.toHexString(0xff & hash[i]);
+        if(hex.length() == 1) hexString.append('0');
```

Comment:
Please correct the indentation. I had to look several times at the code to identify correctly, that line 131 is not executed based on the `if` on line 130...